`typing.TypedDict` is not supported in with pydantic in Python < 3.12

### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the [LangGraph](https://langchain-ai.github.io/langgraph/)/LangChain documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangGraph/LangChain rather than my code.
- [X] I am sure this is better as an issue [rather than a GitHub discussion](https://github.com/langchain-ai/langgraph/discussions/new/choose), since this is a LangGraph bug and not a design question.

### Example Code

```python
# Define agent following the tutorial
from typing import Literal
from langchain_core.runnables import Runnable
from langchain_core.tools import tool
from langchain_openai import ChatOpenAI
from langgraph.prebuilt import create_react_agent

model = ChatOpenAI(model="gpt-4o", temperature=0)

@tool
def get_weather(city: Literal["nyc", "sf"]):
    """Use this to get weather information."""
    if city == "nyc":
        return "It might be cloudy in nyc"
    elif city == "sf":
        return "It's always sunny in sf"
    else:
        raise AssertionError("Unknown city")

tools = [get_weather]
agent = create_react_agent(model, tools=tools)


# Function for traversing the nodes in graph
def traverse_node(node):
    yield node

    if isinstance(node, Runnable):
        nodes = node.get_graph().nodes.values()
        for node in nodes:
            yield from traverse_node(node.data)

list(traverse_node(agent))
```


### Error Message and Stack Trace (if applicable)

```shell
tests/langchain/test_traversal.py:31: in traverse_node
    yield from traverse_node(node.data)
        Runnable   = <class 'langchain_core.runnables.base.Runnable'>
        node       = Node(id='agent', name='agent', data=RunnableLambda(call_model), metadata=None)
        nodes      = dict_values([Node(id='__start__', name='__start__', data=<class 'langchain_core.utils.pydantic.LangGraphInput'>, metad...one), Node(id='__end__', name='__end__', data=<class 'langchain_core.utils.pydantic.LangGraphOutput'>, metadata=None)])
        traverse_node = <function test_hoge.<locals>.traverse_node at 0x7f8c09212af0>
tests/langchain/test_traversal.py:27: in traverse_node
    nodes = node.get_graph().nodes.values()
        Runnable   = <class 'langchain_core.runnables.base.Runnable'>
        node       = RunnableLambda(call_model)
        traverse_node = <function test_hoge.<locals>.traverse_node at 0x7f8c09212af0>
.venv/dev/lib/python3.9/site-packages/langchain_core/runnables/base.py:4495: in get_graph
    input_node = graph.add_node(self.get_input_schema(config))
        __class__  = <class 'langchain_core.runnables.base.RunnableLambda'>
        config     = None
        deps       = [RunnableLambda(...)
| RunnableBinding(bound=ChatOpenAI(client=<openai.resources.chat.completions.Completions object a...enum': ['nyc', 'sf'], 'type': 'string'}}, 'required': ['city'], 'type': 'object'}}}]}, config={}, config_factories=[])]
        graph      = Graph(nodes={}, edges=[])
        self       = RunnableLambda(call_model)
.venv/dev/lib/python3.9/site-packages/langchain_core/runnables/base.py:4403: in get_input_schema
    return super().get_input_schema(config)
        __class__  = <class 'langchain_core.runnables.base.RunnableLambda'>
        config     = None
        func       = <function create_react_agent.<locals>.call_model at 0x7f8c0927fe50>
        self       = RunnableLambda(call_model)
.venv/dev/lib/python3.9/site-packages/langchain_core/runnables/base.py:351: in get_input_schema
    return create_model_v2(
        config     = None
        root_type  = <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>
        self       = RunnableLambda(call_model)
.venv/dev/lib/python3.9/site-packages/langchain_core/utils/pydantic.py:594: in create_model_v2
    named_root_model = _create_root_model(
        field_definitions = {}
        kwargs     = {'type_': <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>}
        model_name = 'call_model_input'
        module_name = 'langchain_core.runnables.base'
        root       = <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>
.venv/dev/lib/python3.9/site-packages/langchain_core/utils/pydantic.py:449: in _create_root_model
    custom_root_type = type(name, (RootModel,), base_class_attributes)
        base_class_attributes = {'__annotations__': {'root': <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>}, '__class_vars__': set(), '__module__': 'langchain_core.runnables.base', '__private_attributes__': {}, ...}
        default_   = <object object at 0x7f8b58e8d9a0>
        model_json_schema = <function _create_root_model.<locals>.model_json_schema at 0x7f8c09170ee0>
        module_name = 'langchain_core.runnables.base'
        name       = 'call_model_input'
        schema     = <function _create_root_model.<locals>.schema at 0x7f8c09170e50>
        type_      = <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>
.venv/dev/lib/python3.9/site-packages/pydantic/_internal/_model_construction.py:224: in __new__
    complete_model_class(
        BaseModel  = <class 'pydantic.main.BaseModel'>
        __class__  = <class 'pydantic._internal._model_construction.ModelMetaclass'>
        __pydantic_generic_metadata__ = None
        __pydantic_reset_parent_namespace__ = True
        _create_model_module = None
        base_field_names = {'root'}
        base_private_attributes = {}
        bases      = (<class 'pydantic.root_model.RootModel'>,)
        class_vars = set()
        cls        = <class 'langchain_core.runnables.base.call_model_input'>
        cls_name   = 'call_model_input'
        config_wrapper = ConfigWrapper(arbitrary_types_allowed=True)
        kwargs     = {}
        mcs        = <class 'pydantic._internal._model_construction.ModelMetaclass'>
        mro        = (<class 'langchain_core.runnables.base.call_model_input'>, <class 'pydantic.root_model.RootModel'>, <class 'pydantic.main.BaseModel'>, <class 'typing.Generic'>, <class 'object'>)
        namespace  = {'__annotations__': {'root': <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>}, '__class_vars__': set(), '__module__': 'langchain_core.runnables.base', '__private_attributes__': {}, ...}
        parameters = (~RootModelRootType,)
        parent_namespace = {'base_class_attributes': {'__annotations__': {'root': <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>}, '... _create_root_model.<locals>.model_json_schema at 0x7f8c09170ee0>, 'module_name': 'langchain_core.runnables.base', ...}
        parent_parameters = (~RootModelRootType,)
        private_attributes = {}
        types_namespace = {'ABC': <class 'abc.ABC'>, 'AddableDict': <class 'langchain_core.runnables.utils.AddableDict'>, 'Any': typing.Any, 'An..._core.runnables.utils.ConfigurableFieldSingleOption, langchain_core.runnables.utils.ConfigurableFieldMultiOption], ...}
.venv/mlflow-remote/lib/python3.9/site-packages/pydantic/_internal/_model_construction.py:577: in complete_model_class
    schema = cls.__get_pydantic_core_schema__(cls, handler)
        cls        = <class 'langchain_core.runnables.base.call_model_input'>
        cls_name   = 'call_model_input'
        config_wrapper = ConfigWrapper(arbitrary_types_allowed=True)
        create_model_module = None
        gen_schema = <pydantic._internal._generate_schema.GenerateSchema object at 0x7f8c09119950>
        handler    = <pydantic._internal._schema_generation_shared.CallbackGetCoreSchemaHandler object at 0x7f8c0915ef40>
        raise_errors = False
        types_namespace = {'ABC': <class 'abc.ABC'>, 'AddableDict': <class 'langchain_core.runnables.utils.AddableDict'>, 'Any': typing.Any, 'An..._core.runnables.utils.ConfigurableFieldSingleOption, langchain_core.runnables.utils.ConfigurableFieldMultiOption], ...}
        typevars_map = {}
.venv/mlflow-remote/lib/python3.9/site-packages/pydantic/main.py:671: in __get_pydantic_core_schema__
    return handler(source)
        cls        = <class 'langchain_core.runnables.base.call_model_input'>
        handler    = <pydantic._internal._schema_generation_shared.CallbackGetCoreSchemaHandler object at 0x7f8c0915ef40>
        schema     = None
        source     = <class 'langchain_core.runnables.base.call_model_input'>
.venv/dev/lib/python3.9/site-packages/pydantic/_internal/_schema_generation_shared.py:83: in __call__
    schema = self._handler(source_type)
        self       = <pydantic._internal._schema_generation_shared.CallbackGetCoreSchemaHandler object at 0x7f8c0915ef40>
        source_type = <class 'langchain_core.runnables.base.call_model_input'>
.venv/dev/lib/python3.9/site-packages/pydantic/_internal/_generate_schema.py:655: in generate_schema
    schema = self._generate_schema_inner(obj)
        from_dunder_get_core_schema = False
        obj        = <class 'langchain_core.runnables.base.call_model_input'>
        schema     = None
        self       = <pydantic._internal._generate_schema.GenerateSchema object at 0x7f8c09119950>
.venv/dev/lib/python3.9/site-packages/pydantic/_internal/_generate_schema.py:924: in _generate_schema_inner
    return self._model_schema(obj)
        BaseModel  = <class 'pydantic.main.BaseModel'>
        obj        = <class 'langchain_core.runnables.base.call_model_input'>
        self       = <pydantic._internal._generate_schema.GenerateSchema object at 0x7f8c09119950>
.venv/dev/lib/python3.9/site-packages/pydantic/_internal/_generate_schema.py:724: in _model_schema
    root_field = self._common_field_schema('root', fields['root'], decorators)
        cls        = <class 'langchain_core.runnables.base.call_model_input'>
        computed_fields = {}
        config_wrapper = ConfigWrapper(arbitrary_types_allowed=True)
        core_config = {'title': 'call_model_input'}
        decorators = DecoratorInfos(validators={}, field_validators={}, root_validators={}, field_serializers={}, model_serializers={}, model_validators={}, computed_fields={})
        extras_schema = None
        fields     = {'root': FieldInfo(annotation=AgentState, required=True)}
        maybe_schema = None
        metadata   = {'pydantic_js_annotation_functions': [], 'pydantic_js_functions': [functools.partial(<function modify_model_json_schema at 0x7f8b59acf700>, cls=<class 'langchain_core.runnables.base.call_model_input'>, title=None)]}
        model_ref  = 'langchain_core.runnables.base.call_model_input:94900564014896'
        model_validators = dict_values([])
        self       = <pydantic._internal._generate_schema.GenerateSchema object at 0x7f8c09119b80>
        title      = None
.venv/dev/lib/python3.9/site-packages/pydantic/_internal/_generate_schema.py:1308: in _common_field_schema
    schema = self._apply_annotations(
        FieldInfo  = <class 'pydantic.fields.FieldInfo'>
        annotations = []
        decorators = DecoratorInfos(validators={}, field_validators={}, root_validators={}, field_serializers={}, model_serializers={}, model_validators={}, computed_fields={})
        field_info = FieldInfo(annotation=AgentState, required=True)
        name       = 'root'
        self       = <pydantic._internal._generate_schema.GenerateSchema object at 0x7f8c09119b80>
        set_discriminator = <function GenerateSchema._common_field_schema.<locals>.set_discriminator at 0x7f8c09170f70>
        source_type = <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>
        validators_from_decorators = []
.venv/dev/lib/python3.9/site-packages/pydantic/_internal/_generate_schema.py:2107: in _apply_annotations
    schema = get_inner_schema(source_type)
        annotations = []
        get_inner_schema = <pydantic._internal._schema_generation_shared.CallbackGetCoreSchemaHandler object at 0x7f8c09280fd0>
        inner_handler = <function GenerateSchema._apply_annotations.<locals>.inner_handler at 0x7f8c0911a040>
        pydantic_js_annotation_functions = []
        res        = None
        self       = <pydantic._internal._generate_schema.GenerateSchema object at 0x7f8c09119b80>
        source_type = <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>
        transform_inner_schema = <function GenerateSchema.<lambda> at 0x7f8b59aded30>
.venv/dev/lib/python3.9/site-packages/pydantic/_internal/_schema_generation_shared.py:83: in __call__
    schema = self._handler(source_type)
        self       = <pydantic._internal._schema_generation_shared.CallbackGetCoreSchemaHandler object at 0x7f8c09280fd0>
        source_type = <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>
.venv/dev/lib/python3.9/site-packages/pydantic/_internal/_generate_schema.py:2088: in inner_handler
    schema = self._generate_schema_inner(obj)
        from_property = None
        obj        = <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>
        self       = <pydantic._internal._generate_schema.GenerateSchema object at 0x7f8c09119b80>
        source_type = <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>
        transform_inner_schema = <function GenerateSchema.<lambda> at 0x7f8b59aded30>
.venv/dev/lib/python3.9/site-packages/pydantic/_internal/_generate_schema.py:929: in _generate_schema_inner
    return self.match_type(obj)
        BaseModel  = <class 'pydantic.main.BaseModel'>
        obj        = <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>
        self       = <pydantic._internal._generate_schema.GenerateSchema object at 0x7f8c09119b80>
.venv/dev/lib/python3.9/site-packages/pydantic/_internal/_generate_schema.py:999: in match_type
    return self._typed_dict_schema(obj, None)
        obj        = <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>
        self       = <pydantic._internal._generate_schema.GenerateSchema object at 0x7f8c09119b80>
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pydantic._internal._generate_schema.GenerateSchema object at 0x7f8c09119b80>
typed_dict_cls = <class 'langgraph.prebuilt.chat_agent_executor.AgentState'>
origin = None

    def _typed_dict_schema(self, typed_dict_cls: Any, origin: Any) -> core_schema.CoreSchema:
        """Generate schema for a TypedDict.
    
        It is not possible to track required/optional keys in TypedDict without __required_keys__
        since TypedDict.__new__ erases the base classes (it replaces them with just `dict`)
        and thus we can track usage of total=True/False
        __required_keys__ was added in Python 3.9
        (https://github.com/miss-islington/cpython/blob/1e9939657dd1f8eb9f596f77c1084d2d351172fc/Doc/library/typing.rst?plain=1#L1546-L1548)
        however it is buggy
        (https://github.com/python/typing_extensions/blob/ac52ac5f2cb0e00e7988bae1e2a1b8257ac88d6d/src/typing_extensions.py#L657-L666).
    
        On 3.11 but < 3.12 TypedDict does not preserve inheritance information.
    
        Hence to avoid creating validators that do not do what users expect we only
        support typing.TypedDict on Python >= 3.12 or typing_extension.TypedDict on all versions
        """
        FieldInfo = import_cached_field_info()
    
        with self.model_type_stack.push(typed_dict_cls), self.defs.get_schema_or_ref(typed_dict_cls) as (
            typed_dict_ref,
            maybe_schema,
        ):
            if maybe_schema is not None:
                return maybe_schema
    
            typevars_map = get_standard_typevars_map(typed_dict_cls)
            if origin is not None:
                typed_dict_cls = origin
    
            if not _SUPPORTS_TYPEDDICT and type(typed_dict_cls).__module__ == 'typing':
>               raise PydanticUserError(
                    'Please use `typing_extensions.TypedDict` instead of `typing.TypedDict` on Python < 3.12.',
                    code='typed-dict-version',
                )
E               pydantic.errors.PydanticUserError: Please use `typing_extensions.TypedDict` instead of `typing.TypedDict` on Python < 3.12.
E               
E               For further information visit https://errors.pydantic.dev/2.9/u/typed-dict-version
```


### Description

The prebuilt ReAct agent code uses `typing.TypedDict` ([code](https://github.com/langchain-ai/langgraph/blob/942dfa841e18760c9a2b85740390ba84cff8ca84/libs/langgraph/langgraph/graph/message.py#L2)), which is not supported in pydantic v2 with Python < 3.12.
https://docs.pydantic.dev/2.9/errors/usage_errors/#typed-dict-version

This causes several errors, for example, calling `get_graph()` of subnode raises the pydantic user error above. To make it compatible with Python < 3.12, the import should be replaced with `typing_extensions.TypedDict` as pydantic recommends.


### System Info

System Information
------------------
> OS:  Linux
> OS Version:  #134+fips1-Ubuntu SMP Wed Apr 10 14:20:32 UTC 2024
> Python Version:  3.9.6 (default, Jul  9 2024, 14:21:06) 
[GCC 9.4.0]

Package Information
-------------------
> langchain_core: 0.3.5
> langchain: 0.3.0
> langsmith: 0.1.125
> langchain_experimental: 0.3.1
> langchain_openai: 0.2.0
> langchain_text_splitters: 0.3.0
> langgraph: 0.2.19

Optional packages not installed
-------------------------------
> langserve

Other Dependencies
------------------
> aiohttp: 3.9.5
> async-timeout: 4.0.3
> httpx: 0.27.0
> jsonpatch: 1.33
> langchain-community: Installed. No version info available.
> langgraph-checkpoint: 1.0.3
> numpy: 1.26.4
> openai: 1.42.0
> orjson: 3.10.6
> packaging: 23.2
> pydantic: 2.9.2
> PyYAML: 6.0.1
> requests: 2.32.3
> SQLAlchemy: 2.0.31
> tenacity: 8.5.0
> tiktoken: 0.7.0
> typing-extensions: 4.12.2

--- a/libs/langgraph/langgraph/prebuilt/chat_agent_executor.py
+++ b/libs/langgraph/langgraph/prebuilt/chat_agent_executor.py
@@ -1,19 +1,10 @@
-from typing import (
-    Annotated,
-    Callable,
-    Literal,
-    Optional,
-    Sequence,
-    Type,
-    TypedDict,
-    TypeVar,
-    Union,
-)
+from typing import Callable, Literal, Optional, Sequence, Type, TypeVar, Union
 
 from langchain_core.language_models import BaseChatModel
 from langchain_core.messages import AIMessage, BaseMessage, SystemMessage, ToolMessage
 from langchain_core.runnables import Runnable, RunnableConfig, RunnableLambda
 from langchain_core.tools import BaseTool
+from typing_extensions import Annotated, TypedDict
 
 from langgraph._api.deprecation import deprecated_parameter
 from langgraph.graph import StateGraph

